File: packages/vault-item-types/src/schema.ts
/**
 * The `VaultItemType` manifest (ADR 0087 §1).
 *
 * A definition is inert data: it names field types from the closed catalogue,
 * declares how the type projects onto the base native secret, and says which
 * of its fields are safe to render without a reveal gesture. It carries no
 * value, no URL, and no code. Everything a client needs to draw the ceremony
 * and read the item back is here.
 */

import type { FieldTypeId } from "./catalogue.js";

export const DEFINITION_API_VERSION = "opensesame.dev/v1alpha1";
export const DEFINITION_KIND = "VaultItemType";
/** The publisher only builtin definitions may claim (ADR 0087 §5). */
export const PLATFORM_PUBLISHER = "https://opensesame.dev";

export const MAX_DEFINITION_BYTES = 64 * 1024;
export const MAX_SECTIONS = 16;
export const MAX_FIELDS = 64;
export const MAX_OPTIONS = 32;
export const MAX_LABEL_CHARS = 120;

/**
 * Ceremonies the platform implements because they do something no data
 * description can: issue a certificate against the Host API, open a drop
 * claim session, classify passkey custody, bound a later delegation. Only a
 * platform-published definition may name one (ADR 0087 §6).
 */
export const HANDLER_IDS = [
  "login",
  "passkey",
  "secret",
  "certificate",
  "drop",
] as const;

export type HandlerId = (typeof HANDLER_IDS)[number];

/**
 * FIDO CXF credential discriminators a definition may map onto. `custom-fields`
 * is the floor: it is what CXF defines for what the standard did not
 * anticipate, so no community type is unexportable (ADR 0087 §4).
 */
export const CXF_CREDENTIAL_IDS = [
  "basic-auth",
  "passkey",
  "totp",
  "credit-card",
  "note",
  "ssh-key",
  "api-key",
  "wifi",
  "address",
  "person-name",
  "identity-document",
  "drivers-license",
  "passport",
  "file",
  "custom-fields",
] as const;

export type CxfCredentialId = (typeof CXF_CREDENTIAL_IDS)[number];

export type FieldDefinition = {
  readonly id: string;
  readonly type: FieldTypeId;
  readonly label: string;
  readonly help?: string;
  readonly required?: boolean;
  readonly placeholder?: string;
  /** `select` only; the complete closed option list. */
  readonly options?: readonly string[];
  /** Scalar shapes only; the field holds an ordered list of values. */
  readonly multiple?: boolean;
  /** Prefilled in a new item. Refused outright on a concealed field type. */
  readonly default?: string;
};

export type SectionDefinition = {
  readonly id: string;
  readonly title: string;
  readonly fields: readonly FieldDefinition[];
};

export type TrailerMapping = {
  /** The `key:` written into the native entry's trailer. */
  readonly key: string;
  /** The field whose value it carries. */
  readonly field: string;
};

/**
 * How the type projects onto `sealed_store::Entry`: one field becomes line
 * one, the rest become trailer keys. `secret: null` is a type with nothing on
 * line one (a note, an address), which projects to an empty first line.
 */
export type NativeProjection = {
  readonly secret: string | null;
  readonly trailer: readonly TrailerMapping[];
};

export type CxfMapping = {
  readonly credential: CxfCredentialId;
};

export type ItemTypeSpec = {
  readonly title: string;
  readonly plural: string;
  /** VFS filename extension, leading dot (ADR 0064/0073). */
  readonly extension: string;
  readonly summary: string;
  readonly categories: readonly string[];
  readonly sections: readonly SectionDefinition[];
  readonly native: NativeProjection;
  readonly cxf: CxfMapping;
  /** Fields shown in the item list. Never concealed ones (ADR 0087 §5). */
  readonly subtitle: readonly string[];
  /** Fields added to the search haystack. Never concealed ones. */
  readonly search: readonly string[];
  /** Platform-published definitions only. */
  readonly handler?: HandlerId;
};

export type ItemTypeMetadata = {
  readonly id: string;
  readonly version: string;
  readonly publisher: string;
};

export type ItemTypeDefinition = {
  readonly apiVersion: typeof DEFINITION_API_VERSION;
  readonly kind: typeof DEFINITION_KIND;
  readonly metadata: ItemTypeMetadata;
  readonly spec: ItemTypeSpec;
};

/**
 * A field value as stored in the vault body: text, a list of text for a
 * repeating field, or the named parts of a record-shaped one. Text throughout,
 * because the base native secret is text and every type projects onto it.
 */
export type FieldValue = string | string[] | Record<string, string>;

export type FieldValues = Readonly<Record<string, FieldValue>>;

/** Every field of a definition, flattened in section then declaration order. */
export function definitionFields(
  definition: ItemTypeDefinition,
): readonly FieldDefinition[] {
  return definition.spec.sections.flatMap((section) => section.fields);
}

export function findField(
  definition: ItemTypeDefinition,
  fieldId: string,
): FieldDefinition | undefined {
  return definitionFields(definition).find((field) => field.id === fieldId);
}

export function isPlatformPublished(definition: ItemTypeDefinition): boolean {
  return definition.metadata.publisher === PLATFORM_PUBLISHER;
}

File: crates/vault-item-types/src/schema.rs
//! The `VaultItemType` manifest, as the host plane parses it (ADR 0087 §1).
//!
//! `deny_unknown_fields` on every struct, for the reason ADR 0065 §4 gives: an
//! author must not be able to smuggle in a field this parser silently drops
//! today and some later consumer honours. There is no field here capable of
//! carrying a URL a loader would fetch, and none capable of carrying a value —
//! `default` is refused on every concealed field type by `validate`.

use serde::{Deserialize, Serialize};

use crate::catalogue::FieldTypeId;

pub const DEFINITION_API_VERSION: &str = "opensesame.dev/v1alpha1";
pub const DEFINITION_KIND: &str = "VaultItemType";
/// The publisher only built-in definitions may claim (ADR 0087 §5).
pub const PLATFORM_PUBLISHER: &str = "https://opensesame.dev";

pub const MAX_DEFINITION_BYTES: usize = 64 * 1024;
pub const MAX_SECTIONS: usize = 16;
pub const MAX_FIELDS: usize = 64;
pub const MAX_OPTIONS: usize = 32;
pub const MAX_LABEL_CHARS: usize = 120;
pub const MAX_SUMMARY_CHARS: usize = 400;

/// Ceremonies the platform implements because no data description can express
/// them. Only a platform-published definition may name one (ADR 0087 §6).
#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum HandlerId {
    Login,
    Passkey,
    Secret,
    Certificate,
    Drop,
}

/// FIDO CXF credential discriminators a definition may map onto.
/// `CustomFields` is the floor, so nothing is unexportable (ADR 0087 §4).
#[derive(Clone, Copy, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum CxfCredentialId {
    BasicAuth,
    Passkey,
    Totp,
    CreditCard,
    Note,
    SshKey,
    ApiKey,
    Wifi,
    Address,
    PersonName,
    IdentityDocument,
    DriversLicense,
    Passport,
    File,
    CustomFields,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct FieldDefinition {
    pub id: String,
    #[serde(rename = "type")]
    pub field_type: FieldTypeId,
    pub label: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub help: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub required: Option<bool>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub placeholder: Option<String>,
    /// `select` only; the complete closed option list.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub options: Option<Vec<String>>,
    /// Scalar shapes only; the field holds an ordered list of values.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub multiple: Option<bool>,
    /// Prefilled in a new item. Refused outright on a concealed field type.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub default: Option<String>,
}

impl FieldDefinition {
    #[must_use]
    pub fn repeats(&self) -> bool {
        self.multiple == Some(true)
    }

    #[must_use]
    pub fn is_required(&self) -> bool {
        self.required == Some(true)
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct SectionDefinition {
    pub id: String,
    pub title: String,
    pub fields: Vec<FieldDefinition>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct TrailerMapping {
    /// The `key:` written into the native entry's trailer.
    pub key: String,
    /// The field whose value it carries.
    pub field: String,
}

/// How the type lands on `sealed_store::Entry`: one field becomes line one,
/// the rest become trailer keys. `secret: null` projects an empty first line.
#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct NativeProjection {
    pub secret: Option<String>,
    pub trailer: Vec<TrailerMapping>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct CxfMapping {
    pub credential: CxfCredentialId,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct ItemTypeSpec {
    pub title: String,
    pub plural: String,
    /// VFS filename extension, leading dot (ADR 0064/0073).
    pub extension: String,
    pub summary: String,
    pub categories: Vec<String>,
    pub sections: Vec<SectionDefinition>,
    pub native: NativeProjection,
    pub cxf: CxfMapping,
    /// Fields shown in the item list. Never concealed ones (ADR 0087 §5).
    pub subtitle: Vec<String>,
    /// Fields added to the search haystack. Never concealed ones.
    pub search: Vec<String>,
    /// Platform-published definitions only.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub handler: Option<HandlerId>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields)]
pub struct ItemTypeMetadata {
    pub id: String,
    pub version: String,
    pub publisher: String,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq, Eq)]
#[serde(deny_unknown_fields, rename_all = "camelCase")]
pub struct ItemTypeDefinition {
    pub api_version: String,
    pub kind: String,
    pub metadata: ItemTypeMetadata,
    pub spec: ItemTypeSpec,
}

impl ItemTypeDefinition {
    /// Every field, flattened in section then declaration order.
    #[must_use]
    pub fn fields(&self) -> Vec<&FieldDefinition> {
        self.spec
            .sections
            .iter()
            .flat_map(|section| section.fields.iter())
            .collect()
    }

    #[must_use]
    pub fn field(&self, id: &str) -> Option<&FieldDefinition> {
        self.fields().into_iter().find(|field| field.id == id)
    }

    #[must_use]
    pub fn is_platform_published(&self) -> bool {
        self.metadata.publisher == PLATFORM_PUBLISHER
    }
}
